### PR TITLE
Changes to rnn.py for multitask compatibility

### PR DESCRIPTION
**Pull request by borguz**
Summary:
Some changes to rnn.py to better support multitasking:
- Add mean and max pooling as alternative to  dot attention
- Allow removing attention alltogether
- Padding value parameter

Reviewed By: liezl200

Differential Revision: D8062147

fbshipit-source-id: 59c6d95c3812cc79f8fcb05a3526846d347d6f72

### CODE REVIEW

File: pytorch_translate/common_layers.py [hunk 1]
```diff
@@ -49,10 +50,57 @@ def forward(self, decoder_state, source_hids, src_lengths):
...
+                source_hids,
+                src_lengths,
+            )
         else:
             raise ValueError(f"Attention type {self.attention_type} is not supported")
         return output, attn_scores
 
+    def pooling(self, pool_type, source_hids, src_lengths):
+        assert self.decoder_hidden_state_dim == self.encoder_output_dim
+        max_srclen = source_hids.size()[0]
```
Comment (jmp84) [inline pytorch_translate/common_layers.py]: `max_src_len`

File: pytorch_translate/common_layers.py [hunk 2]
```diff
@@ -49,10 +50,57 @@ def forward(self, decoder_state, source_hids, src_lengths):
...
+        denom = src_lengths.t().unsqueeze(2).type_as(source_hids)
+        if pool_type == "mean":
+            masked_hiddens = source_hids * src_mask
+            context = (masked_hiddens / denom).sum(dim=0)
+        elif pool_type == "max":
+            masked_hiddens = source_hids - 10e6 * (1 - src_mask)
+            context = masked_hiddens.max(dim=0)
+        else:
+            raise ValueError(
+                'Pooling type {} is not supported'.format(pool_type)
```
Comment (jmp84) [inline pytorch_translate/common_layers.py]: `f'Pooling type {pool_type} is not supported'`

File: pytorch_translate/common_layers.py [hunk 3]
```diff
@@ -49,10 +50,57 @@ def forward(self, decoder_state, source_hids, src_lengths):
...
+                source_hids.data,
+            ),
+            requires_grad=False,
+        ).t().unsqueeze(2)
+        denom = src_lengths.t().unsqueeze(2).type_as(source_hids)
+        if pool_type == "mean":
+            masked_hiddens = source_hids * src_mask
+            context = (masked_hiddens / denom).sum(dim=0)
+        elif pool_type == "max":
+            masked_hiddens = source_hids - 10e6 * (1 - src_mask)
```
Comment (jmp84) [inline pytorch_translate/common_layers.py]: How do you choose this constant?

Comment (borguz) [reply pytorch_translate/common_layers.py]: This serves as a replacement for -infinity.  Anything larger than typical hidden outputs would work.  The reason we did not use -infinity is because it causes NaN's downstream which can break the rest of the code (implementation details).
